diff --git a/airflow/www/templates/airflow/graph.html b/airflow/www/templates/airflow/graph.html
index 24fc508..1bf2740 100644
--- a/airflow/www/templates/airflow/graph.html
+++ b/airflow/www/templates/airflow/graph.html
@@ -1,13 +1,13 @@
-{# 
+{#
   Licensed to the Apache Software Foundation (ASF) under one or more
   contributor license agreements.  See the NOTICE file distributed with
   this work for additional information regarding copyright ownership.
   The ASF licenses this file to You under the Apache License, Version 2.0
   (the "License"); you may not use this file except in compliance with
   the License.  You may obtain a copy of the License at
-  
+
     http://www.apache.org/licenses/LICENSE-2.0
-  
+
   Unless required by applicable law or agreed to in writing, software
   distributed under the License is distributed on an "AS IS" BASIS,
   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
@@ -54,6 +54,12 @@
     </div>
 <hr/>
 <div>
+    {% if dr_conf %}
+    Conf:
+    <pre>
+    {{ dr_conf }}
+    </pre>
+    {% endif %}
     {% for op in operators %}
     <div class="legend_item" style="border-width:1px;float:left;background:{{ op.ui_color }};color:{{ op.ui_fgcolor }};">
         {{ op.__name__ }}
diff --git a/airflow/www/views.py b/airflow/www/views.py
index 3788276..9a2f490 100644
--- a/airflow/www/views.py
+++ b/airflow/www/views.py
@@ -23,6 +23,7 @@ from datetime import datetime, timedelta
 import dateutil.parser
 import copy
 import json
+import pprint
 
 import inspect
 from textwrap import dedent
@@ -1321,10 +1322,12 @@ class Airflow(BaseView):
         )
         dr_choices = []
         dr_state = None
+        dr_conf = None
         for dr in drs:
             dr_choices.append((dr.execution_date.isoformat(), dr.run_id))
             if dttm == dr.execution_date:
                 dr_state = dr.state
+                dr_conf = dr.conf
 
         class GraphForm(Form):
             execution_date = SelectField("DAG run", choices=dr_choices)
@@ -1351,6 +1354,7 @@ class Airflow(BaseView):
         session.commit()
         session.close()
         doc_md = markdown.markdown(dag.doc_md) if hasattr(dag, 'doc_md') and dag.doc_md else ''
+        dr_conf = pprint.pformat(dr_conf, indent=4) if dr_conf else ''
 
         return self.render(
             'airflow/graph.html',
@@ -1361,6 +1365,7 @@ class Airflow(BaseView):
             execution_date=dttm.isoformat(),
             state_token=state_token(dr_state),
             doc_md=doc_md,
+            dr_conf=dr_conf,
             arrange=arrange,
             operators=sorted(
                 list(set([op.__class__ for op in dag.tasks])),
